Install StorageOS v2 with Etcd (#31)
storageos/tutorials: install-storageos/040-storage-class.md

Lets check how to use dynamically provisioned storage using
StorageOS.

> StorageOS creates a StorageClass by default, called `fast`. Many more can be

Reviewer: ```suggestion
StorageOS creates a StorageClass by default, called `fast`.  Other StorageClasses can be 
```

> created to define different behaviour or concern.

Reviewer: ```suggestion
created to define different behaviours or apply feature labels by default.
```

```
kubectl create -f-<<END
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: dev
provisioner: csi.storageos.com
parameters:
  csi.storage.k8s.io/fstype: ext4
  pool: default

  # Change the Namespace below if StorageOS doesn't run in kube-system
  csi.storage.k8s.io/node-publish-secret-namespace: kube-system
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-node-publish-secret
  csi.storage.k8s.io/provisioner-secret-name: csi-provisioner-secret
  csi.storage.k8s.io/controller-publish-secret-name: csi-controller-publish-secret
END
```{{execute}}

And a StorageClass for production. Note that we are enabling any volume using
the prod StorageClass to have one replica. Hence, 2 synchronous copies of the
data. The primary volume, where I/O operations are performed by the application
and the replica that keeps in sync from the primary.

```
kubectl create -f- <<END
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: prod
provisioner: csi.storageos.com
parameters:
  csi.storage.k8s.io/fstype: ext4
  pool: default

>   storageos.com/replicas: "1" # Enforces 1 replica for the Volume

Reviewer: ```suggestion
  storageos.com/replicas: "1" # Enforces 1 replica for any Volume created using this StorageClass
```

Lets check how to dynamically provision storage using
StorageOS.

During cluster bootstrap StorageOS creates a StorageClass called `fast`. A
StorageClass is a Kubernetes construct that describes different "classes" of
Storage and their providers.

Additional StorageClasses can be created to define different behaviours or apply
feature labels by default.

Lets create two additional StorageClasses for a practical example of how StorageClasses can be used.

Start by creating a StorageClass for development.

```
kubectl create -f-<<END
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: dev
provisioner: csi.storageos.com
parameters:
  csi.storage.k8s.io/fstype: ext4
  pool: default

> 

Reviewer 2: remove line

  csi.storage.k8s.io/node-publish-secret-namespace: kube-system
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/controller-expand-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-node-publish-secret
  csi.storage.k8s.io/provisioner-secret-name: csi-provisioner-secret
  csi.storage.k8s.io/controller-publish-secret-name: csi-controller-publish-secret
  csi.storage.k8s.io/controller-expand-secret-name: csi-controller-expand-secret
END
```{{execute}}

And a StorageClass for production.

```
kubectl create -f- <<END
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: prod
provisioner: csi.storageos.com
parameters:
  csi.storage.k8s.io/fstype: ext4
  pool: default
  storageos.com/replicas: "1" # Enforces 1 replica for any Volume created by this StorageClass

> 

Reviewer 2: remove line here as well
